openlibrary-wikidata-bot/jobs/sync_author_identifiers_from_wikidata.py

```diff
@@ -0,0 +1,210 @@
+    if not("property" in obj):
+         return False
+    if not ("id" in obj["property"]):
+         return False 
+    if key != obj["property"]["id"]:
+         return False
+    if not ("value" in obj):
+        return False
+    if not ("content" in obj["value"]):
+        return False
+    return True
+
+
+def write_error(wd_id, author_key, author_name, error_type, id, details):
+    logger.error(f'{error_type} for {author_key} ({author_name}), wd {wd_id}: {id}: {details}')
+    with open(csv_file_path, mode="a", newline="", encoding="utf-8") as file:
+        writer = csv.writer(file)
+        writer.writerow([wd_id, author_key, author_name, error_type, id, details])
+
+
+def merge_remote_ids(
+    author, incoming_ids, wd_id
+) -> tuple[dict[str, str], int]:
+    """Returns the author's remote IDs merged with a given remote IDs object, as well as a count for how many IDs had conflicts.
+    If incoming_ids is empty, or if there are more conflicts than matches, no merge will be attempted, and the output will be (author.remote_ids, -1).
+    """
+    if not hasattr(author, "remote_ids"):
+        write_error(wd_id, author.olid, author.name, "openlibrary_author_remote_ids_failed_to_fetch", "", "")
+        return {}, -1
+    output = {**author.remote_ids}
+    # Count
+    update_count = 0
+    conflicts = 0
+    for identifier in WD_IDENTIFIERS.values():
+        if identifier in output and identifier in incoming_ids:
+            if output[identifier] != incoming_ids[identifier]:
+                conflicts = conflicts + 1
+                write_error(wd_id, author.olid, author.name, "openlibrary_wikidata_remote_id_collision", identifier, f'{{"ol": "{output[identifier]}", "wd": "{incoming_ids[identifier]}"}}')
+        elif identifier in incoming_ids and identifier not in output:
+            output[identifier] = incoming_ids[identifier]
+            update_count = update_count + 1
+    if conflicts > 0:
+        return author.remote_ids, -1
+    return output, update_count
+
+
+def consolidate_remote_author_ids(sql_path: str, dry_run: bool = True) -> None:
+    ol = OpenLibrary()
+
+    csv.field_size_limit(sys.maxsize)
+
+    # Read in the wikidata DB dump
+    with open(sql_path, mode="r", encoding="utf-8") as file:
+        next(file) # skip the header line
+        reader = csv.reader(file, delimiter="\t")  # Read as TSV
+        for row in reader:
+            if len(row) < 2:
+                continue
+            wikidata_col_raw = row[1]
+            
+            try:
+                parsed_wikidata_json = json.loads(wikidata_col_raw)
+                wd_json_top_level_fields = parsed_wikidata_json["statements"]
+
+                # Skip if no OL ID
+                if not ("P648" in wd_json_top_level_fields):
+                    continue
+
+                # Some of these WD entries match multiple OL authors. 
+                # What to do in that scenario?
+                # For now, flagging that as a problem
```

Comment:
In some of these cases, duplicated/redirected OL authors are marked as deprecated on Wikidata. But I'm not sure how that looks in the API response, so skipping them seems good, and we can improve by analyzing the error log!